ground/src/main/java/com/google/android/ground/ui/map/gms/GoogleMapsFragmentV2.kt

```diff
@@ -0,0 +1,461 @@
+      }
+    }
+    val result = candidates.build()
+    if (!result.isEmpty()) {
+      locationOfInterestClicks.onNext(result)
+    }
+  }
+
+  private fun onMarkerClick(marker: Marker): Boolean =
+    if (getMap().uiSettings.isZoomGesturesEnabled) {
+      markerClicks.onNext(marker.tag as MapPin)
+      // Allow map to pan to marker.
+      false
+    } else {
+      // Prevent map from panning to marker.
+      true
+    }
+
+  override fun getMapPinClicks(): @Hot Observable<MapPin> = markerClicks
+
+  override fun getLocationOfInterestClicks():
+    @Hot Observable<ImmutableList<MapLocationOfInterest>> = locationOfInterestClicks
+
+  override fun getStartDragEvents(): @Hot Flowable<Nil> = startDragEvents
+
+  override fun getCameraMovedEvents(): @Hot Flowable<CameraPosition> = cameraMovedEvents
+
+  override fun getTileProviders(): @Hot Observable<MapBoxOfflineTileProvider> = tileProviders
+
+  override fun getDistanceInPixels(point1: Point, point2: Point): Double {
+    if (map == null) {
+      Timber.e("Null Map reference")
+      return 0.toDouble()
+    }
+    val projection = map!!.projection
+    val loc1 = projection.toScreenLocation(toLatLng(point1))
+    val loc2 = projection.toScreenLocation(toLatLng(point2))
+    val dx = (loc1.x - loc2.x).toDouble()
+    val dy = (loc1.y - loc2.y).toDouble()
+    return sqrt(dx * dx + dy * dy)
+  }
+
+  override fun enableGestures() = getMap().uiSettings.setAllGesturesEnabled(true)
+
+  override fun disableGestures() = getMap().uiSettings.setAllGesturesEnabled(false)
+
+  override fun moveCamera(point: Point) =
+    getMap().moveCamera(CameraUpdateFactory.newLatLng(toLatLng(point)))
+
+  override fun moveCamera(point: Point, zoomLevel: Float) =
+    getMap().moveCamera(CameraUpdateFactory.newLatLngZoom(toLatLng(point), zoomLevel))
+
+  private fun addMapPin(mapPin: MapPin) {
+    val position = toLatLng(mapPin.position)
+    // TODO: add the anchor values into the resource dimensions file
+    val marker =
+      getMap()
+        .addMarker(
+          MarkerOptions()
+            .position(position)
+            .icon(getMarkerIcon(mapPin))
+            .anchor(0.5f, 0.85f)
+            .alpha(1.0f)
+        )
+    markers.add(marker)
+    marker.tag = mapPin
+  }
+
+  private fun getMarkerIcon(mapPin: MapPin): BitmapDescriptor =
+    markerIconFactory.getMarkerIcon(parseColor(mapPin.style.color), currentZoomLevel)
+
+  private fun addMultiPolygon(multiPolygon: MultiPolygon) =
+    multiPolygon.polygons.forEach { addPolygon(it) }
+
+  private fun addPolygon(polygon: Polygon) {
+    // TODO(jsunde): Remove before merging
+    Timber.v("adding polygon with ${polygon.vertices} vertices")
+
+    val options = PolylineOptions()
+    options.clickable(false)
+    val vertices = polygon.vertices.map { point: Point -> toLatLng(point) }.toImmutableList()
+    options.addAll(vertices)
+
+    val polyline = getMap().addPolyline(options)
+    polyline.tag = polygon
```

Comment:
Tag would ultimately be something like `Pair<String, Class<?>>`. For LOIs this would contain the LOI id and LocationOfInterest.class`

Reply:
Done!